### DISCUSSION TITLE
ONVIF support

### DISCUSSION THREAD

**Discussion by zillesta** [scottlamb/moonfire-nvr#131] [General]
Hi,

I've been playing around with moonfire a bit with some cameras and it's really impressive!

I was wondering if some ONVIF support would be out of scope or if it would fit in. I would be interested in trying to experiment a bit with it if that's the case!

**Comment by scottlamb** [comment]
Thanks! It'd likely fit in. ONVIF does a lot of things. What specifically are you thinking of?

A few I've thought about adding with ONVIF:

*   camera discovery—the config tool could suggest "there's a camera at this IP, do you want to add it?"
*   stream URL discovery—it could give you a drop-down of ONVIF profile names as an alternative to specifying a stream URL, and it'd look up the URL dynamically
*   ONVIF analytics events—I actually have some (quick and dirty / prototype-quality) code relating to this in my [playground repository](https://github.com/scottlamb/moonfire-playground).
*   maybe some day, support for configuring the camera's settings from Moonfire's config UI

I found this [Rust onvif library](https://github.com/lumeohq/onvif-rs) a while ago, although it has some problems. One of them is that the code size is *huge* because of the way the XML (de)serialization library it uses works. I made a [proposal](https://github.com/media-io/yaserde/issues/84) to improve that but didn't get any feedback form the library's author, so I haven't taken it further.

**Reply by zillesta** [reply]
I was thinking mainly of camera discovery and getting the stream URLs to start with at least, which seems like nice quality of life stuff that fits into the schema that exists today. 

I also came across the onvif library and yeah the build times are painful. I might take a look at that proposal and see how much I can do/maybe fork the crate. No promises! :)

**Reply by scottlamb** [reply]
Yeah, I'd love to have that. Maybe as an optional feature until the code size gets under control.

**Reply by scottlamb** [reply]
@zillesta Are you still interested in this? I've been playing with a rewrite of `yaserde` based on my proposal. I'd be happy to work together to polish it and implement this kind of functionality with it.
